elide-hbase.md

```diff
@@ -0,0 +1,159 @@
+
+### Entity beans
+In order to support having byte arrays(or any arbitrary primitive types) as column qualifiers and column family names in HBase, some indirection hacks will need to be done with the `@Column` annotation.
+
+The entity bean should contain a map where the key is a String, and value is an Object that represents the actual column family/column qualifier names.
+
+The `name` attribute in `@Column` should look like `"cfPointer1:cqPointer1"`, where `cfPointer1` and `cqPointer1` are keys in the map described above. If keys cannot be found, we consider the name as is.
+
+For example:
+If there's a Person HTable, and looks like:  
+```
+ROW    COLUMN+CELL
+\x01   \x02:\x03, timestamp=1431924360459, value=\x04
+```
+
+Then the HBase entity bean should look like:
+```
+@Entity
+@Table(name = "person")
+@DataStore(cls = HBaseStore.class)
+public class Person {
+  // This map can be injected, doesn't have to be static
+  private final Map<String, Object> columnMap = ImmutableMap.of(
+      "cf1", new byte[]{2}, // Column Family
+      "cq1", new byte[]{3}  // Column Qualifier
+  );
+
+  @Id
+  private byte[] rowKey;
+
+  @Column(name = "timestamp")
+  private Long timestamp;
+
+  @Column(name = "cf1:cq1")
+  private byte[] value;
+
+  @OneToOne/@OneToMany/@ManyToOne/@ManyToMany
+  @DataStore(cls = HibernateStore.class)
+  public Set<HobbyGroup> getHobbyGroups() { return Collections.emptySet(); }
+}
+```
+
+A Relational Database entity bean should look like:
+```
+@Entity
+@Table(name = "hobbyGroup")
+public class HobbyGroup {
+  @Id
+  private long id;
+  private Set<Person> persons;
+
+  @OneToOne/@OneToMany/@ManyToOne/@ManyToMany
+  @DataStore(cls = HBaseStore.class)
+  @RowKeyFn(cls = RowKeyComputer.class)
+  public Set<Person> getPersons() { return Collections.emptySet(); }
+
+  static class RowKeyComputer implements Function<HobbyGroup, Object> { //could return a FilterScope maybe?
+    @Override
+    public Object apply(HobbyGroup hobbyGroup) {
+      byte[] rowKey = new byte[]{hobbyGroup.getId().byteValue()};
+      return rowKey;
+    }
+  }
+}
+```
+
+### Bridging DataStores
+1. Must use a `MultiplexDataStore`
+1. `PersistenceResource` will still be responsible for holding security check logics. `DataStoreTransaction` will have more to do with getting/setting attr, relations, etc. Methods in `DataStoreTransaction` aren't aware of securities at all
+1. Have a custom annotation `@DataStore(cls)`. This will be used to annotate relations of a different data store type in an entity bean.
+```
+@DataStore(cls=HBaseStore.class)
+public Collection<HBaseEntity> getHBaseEntities() { return Collections.emptySet(); }
+```
+1. Add a field `parentTransaction` to `DataStoreTransaction`, so that when using a `MultiplexTransaction`, a detail data store transaction object can get access to its parent(the MultiplexTransaction) and get to its peer(a transaction of a different data store type):
```

Comment:
My original thought is to have a parent transaction reference, so when seeing `@DataStore(cls)` annotation, we can get the transaction of the specific data store type by traversing up to the parent and down to the leaf. But after reading the comment above, `@DataStore` is not necessary, and transaction can be retrieved by looking up in `EntityDictionary` I guess.